walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,136 @@
+        System.out.println("List of attendees in event: ");
+        for (String attendee : event.getAttendees()) {
+          System.out.println(attendee);
+        }
+        // END DELETE
+
+        if (event.getAttendees().contains(person)) {
+          if (!eventTitles.contains(event.getTitle())) {
+            allEventTimes.add(event.getWhen());
+            System.out.println("add event: START: " + event.getWhen().start() + " END: " + event.getWhen().end());
+          }
+        }
+      }
+    }
+
+    // no attendees match any events
+    if (allEventTimes.isEmpty()) {
+      System.out.println("no attendees match events");
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(allEventTimes, TimeRange.ORDER_BY_END); //TO DELETE
+
+
+    System.out.println("sorting complete"); // TO DELETE
+    for (TimeRange aEvent : allEventTimes) {
+      System.out.println("postsort. START: " + aEvent.start() + " END: " + aEvent.end());
+    }
+
+    // Create new time which starts at first opening and ends at  of request
+
+    newTime: 
+    for(TimeRange eventTime : allEventTimes) { 
+      currEnd = eventTime.start();
+      System.out.println("current start: " + currStart + " current end: " + currEnd);
+      TimeRange proposeTime = TimeRange.fromStartEnd(currStart, currEnd, false);
+
+      // check that proposed time is longer than request dureation. Also ensures
+      // that the propsed duration is not negtive
+      if (checkDuration(proposeTime, request.getDuration())) {
+        System.out.println("durration large enough");
+        for (TimeRange anEvent : allEventTimes) {
+          if (anEvent.equals(proposeTime)) {
+            continue;
+          } else if (anEvent.overlaps(proposeTime)) {
+            proposeTime = fixRange(anEvent, proposeTime);
+            if (!checkDuration(proposeTime, request.getDuration())) {
+              currStart = eventTime.end();
+              continue newTime;
+            }
+          }
+        }
+
+        // if the proposed time is already within the collection don't add
+        for (TimeRange RequestOptions : collection) {
+          if (RequestOptions.equals(proposeTime)) {
+            currStart = eventTime.end();
+            continue;
+          }
+        }
+
+        collection.add(proposeTime);
+      }
+      else {
+        System.out.println("duration too short: " + proposeTime.duration());
+      }
+      currStart = eventTime.end();
+    }
+    //  add time gap for end of day, current start is on last 
+    TimeRange proposeTime = TimeRange.fromStartEnd(currStart, DAY_IN_MINS, false);
+    if (proposeTime.duration() > request.getDuration()) {
+      collection.add(proposeTime);
+    }
+
+    return collection;
+    
+  }
+  
+  private boolean checkDuration(TimeRange proposed, long duration) {
+    return proposed.duration() >= duration;
+  }
+
+  private TimeRange fixRange(TimeRange event, TimeRange proposed) {
```

Comment:
Add a function comment.

One particular thing that I noticed is that this doesn't handle when the event contains the proposed range.  Should it?

Reply:
If the event contains the proposed range then the event start is less than the proposed start which is handled in proposed.start > event.start

Reply:
So I am still confused by this function.

The event is the existing time that can't be scheduled over and the proposed is the proposed meeting time.  If the proposed meeting time overlaps the event then we want to adjust it so  that it no longer overlaps.

For this we have five cases.


**case 1:**
```
Event:             [----------]
Proposed:                                [---------]
```
The event and proposed time does not overlap at all. We should return the proposed meeting unchanged.

return proposed.

I don't see where this is handled.

**case 2:**
```
Event:             [----------]
Proposed:               [---------]
```

The proposed meeting overlaps the end of the event.  Here will want to change the start of the proposed meeting to the end of the event.

return [event.end(), proposed.end()]

I think you are trying to handle this in the first if statement  but your returned end time is event.start() which is before the proposed.start() so the meeting doesn't seem valid.

**case 3:**
```
Event:                    [----------]
Proposed:         [---------]
```

The proposed meeting overlaps the start of the event.  Here will want to change the end of the proposed meeting the start of the event..

return [proposed.start(), event.start()]

I think this is what you try to handle in your else if.

**case 4:**
```
Event:            [----------]
Proposed:            [-----]
```
The proposed completely overlaps an existing.  You want to return that there is no possible meeting time available.

You are correct that the if will catch this and return a meeting time with the end before the start which will then be invalid because it will have a negative duration. Though you might want to handle this case explicitly by returning maybe null to indicate that there are no possible meetings.

**case 5:**
```
Event:            [-----]
Proposed:       [----------]       
```
The proposed meeting contains the event.  You  will want to return two possible meeting times.

return [proposed.start(), event.start()]
return [event.end(), proposed.end()]

I don't see this handled.

Reply:
After talking to Brett, I've realized I need to go back and fix the way I write my comments so that they explain whats being modified, what the function requires(or what is assumed when its called) and what variables are affected, so I will do that after I post this. 

For case 1: 
The function is called assuming the event and proposed overlap, so because in case 1 they don't over lap, the case is invalid (will add to my comments)

For Case 2: 
You're correct, I that was the case that I was attempting to fix but I returned the wrong value. I didn't write it out and should have. 

For case 3 that's correct and for case 4 & 5 I see what you mean. I will add more test cases to cover these which I should have done before. I'll get it back soon. 
Thanks!

Reply:
For case 5, I'm not sure how it would be possible to have an event encompassed within the proposed time like that. 

if this is the case:
```
Event:                        [--------]
Proposed:       [--------------------------]  
```
That means there would have been two event on either side of the proposed
```
All Events:   [--]          [------- ]           [--]
Event:                        [--------]
Proposed:       [--------------------------]  
```
but if that's the case, proposed above would never be an option because it would stop at the middle event first.
Because I ordered them by their ending there should never be the case where an event is encompassed by a proposed time. 

I believe this is correct, but I could be overthinking it.